### ISSUE TITLE
Show AltTab on screen with non-dimmed menubar, when the active app has no window

### ISSUE THREAD

**Issue by Reply2Zain** [lwouis/alt-tab-macos#2107]
version: 6.49.0

**Bug**
When focused on a windowless application on an extended display, alt-tab defaults to showing the popup on the main display. Ideally, since focus is on the extended display, alt-tab should display the popup (applications picker) where the focus is (on the extended display). 

**Steps to reproduce the bug**

1. Go to finder on an external display and close the window (focus is on the extended display).
2. Press alt-tab
3. The popup is now on the main display

I have also found this behavior to be the same with other apps such as "Activity Monitor" 

**Specs**
year: 2021
device: MacBook Pro 16"
chip: apple silicon
OS: Ventura (13.0.1)

**Comment by lwouis** [comment]
That's an interesting scenario. However, I'm not sure I understand your setup, and what you would expect.

What preference did you select there?

<img width="286" alt="image" src="https://user-images.githubusercontent.com/106195/201197242-f4f3257a-6c95-4d0a-942f-5efdb4435ade.png">

You say:

> focused on a windowless application on an extended display

> since focus is on the extended display

What do you mean by that. Focus is materialized by a window having keyboard focus. Typically a piece of UI that will react if you press some keys. If there is no window, if the active app (i.e. app with name in the top menubar) has no window, then there is no window catch any keyboard input.

If I picture 2 monitors side-by-side, and the right one is extending the left one, and in that setup, I focus on Finder, then close all its windows. I'm now looking at 2 screens with no window on them. The next keyboard press has nowhere to go. Where should AltTab show? Why right and not left?

I think the only argument for right or left, in the absence of windows, would be if you selected `Screen including the mouse`, and your mouse was on the right monitor. Is that your scenario?

**Comment by Reply2Zain** [comment]
Thanks for the reply. Focus (or keyboard focus) can be on an app without a window. If you just close "finder" then focus is on "finder" despite there not being any UI window active. You can then  press "cmd + n" and then a new window opens. As you're probably aware, you can always see which app is in focus by looking at the name on the top left next to the apple logo.

**Comment by Reply2Zain** [comment]
I tried the setting that you are talking about and making it "screen including mouse" is a good workaround for my use-case but doesn't address the issue and a fix would improve quality of where the application picker shows. If I were to close finder from all displays then the last place where finder was open had the most recent focus. The app name is highlighted on the top left and pressing "cmd + n" opens the new finder window on the focused desktop.

**Comment by lwouis** [comment]
@Reply2Zain I experimented with that stuff. It's a bit more complex than you describe.

Indeed, the user can click on the menubar, and decide which monitor has an "active" menubar (the other monitors then get their menubar dimmed). However, it's not something that reserved for apps-without-any-window.

Scenario: You can have an app that's opened with a window on Display 1. Then you click on the menubar of Display 2. At this point, Display have has its menubar dimmed but it shows the active window. That window has keyboard focus. Yet, Display 2 has the "active" menubar.

In that scenario, I think Display 1 should be considered the "Active screen", since it has an active window. Even though the menubar is dimmed on that display.

**Comment by lwouis** [comment]
@Reply2Zain I experimented with all this today.

First of all, you were not clear about something: you're using this setting: `System Preferences > Desktop & Dock > Mission Control > Display have separate Spaces = ON`. From your initial description, I thought it was OFF. When it's off, AltTab works perfectly in all scenarios I could test.

Now when it's ON, indeed, the user can click on the menubar, and decide which monitor has an "active" menubar (the other monitors then get their menubar dimmed). However, it's not something that reserved for apps-without-any-window.

Scenario: You can have an app that's opened with a window on Display 1. Then you click on the menubar of Display 2. At this point, Display 1 has its menubar dimmed but it shows the active window. That window has keyboard focus. Yet, Display 2 has the "active" menubar.

In that scenario, I think Display 1 should be considered the "Active screen", since it has an active window. Even though the menubar is dimmed on that display. Here is [what Apple calls the "Main display" in its technical docs](https://developer.apple.com/documentation/appkit/nsscreen/1388371-main):

> The main screen is not necessarily the same screen that contains the menu bar or has its origin at (0, 0). The main screen refers to the screen containing the window that is currently receiving keyboard events. It is the main screen because it is the one with which the user is most likely interacting. The screen containing the menu bar is always the first object (index 0) in the array returned by the screens method.

In conclusion, I think AltTab is correct in all scenarios except one where we could improve: if the active app has no window, then we could check which menubar is active to decide which screen is the Active Screen. That being said, I'm not sure at all that there is an API to know which menubar is active. I'll look into it.

**Comment by Reply2Zain** [comment]
I have many of the default settings for macOS Ventura, which is why I didn't get into my specific system settings for desktop.  Also I did try my best to convey as much as I could accurately but it must be still be difficult to follow my thought process on the issue. 

With that said, your commitment to understanding the issue is beyond what I expected. It appears that you fully understood what I was trying to convey, confirmed with your own testing. You identified the specific case where there could be an improvement (when using this on multiple screens) and identified how to fix it as well. This is just me saying thanks for what you've done. I just tried the new update (6.50.0) and the issue is completely resolved. Thank you.
